src/net/socket_state.rs

```diff
@@ -49,19 +49,23 @@ impl SocketState {
         }
 
         let raw_packet: RawPacket;
-        // initialize packet data, seq, acked_seq etc.
-        let mut l = connection
-            .write()
-            .map_err(|_| NetworkError::AddConnectionToManagerFailed)?;
 
-        raw_packet = RawPacket::new(
-            l.seq_num,
-            &packet,
-            l.their_acks.last_seq,
-            l.their_acks.field,
-        );
-        // increase sequence number
-        l.seq_num = l.seq_num.wrapping_add(1);
+        {
+            // initialize packet data, seq, acked_seq etc.
+            let mut l = connection
+                .write()
+                .map_err(|_| NetworkError::AddConnectionToManagerFailed)?;
+
+            raw_packet = RawPacket::new(
```

Comment:
I would propose to grab all the data from your connection (seq_num, acks), increment the sequence number and free the lock. Afterwards, create the raw packet (which includes a payload clone it seems) and serialize it. This should minimize the time the lock will be kept without violating the logic (?)

Reply:
Going to try that since I just tested it further and saw that there still was an problem. Let mee take look here.